### PR TITLE
[wasm] Rationalize wasm jobs run on CI

### PR DESCRIPTION
**Pull request by radical**
Essentially:
- For rolling builds, we want to run all the wasm jobs, unconditionally
   - And `runtime-extra-platforms` runs the tests that were not in `runtime`, and `runtime-staging`
- For PRs, we use the various path checks

Changes:
- split wasm jobs into various templates
- combine, and simplify some jobs

# Updated state of wasm jobs
## PR:
 * `runtime-extra-platforms`, and `runtime-wasm` run only when manually triggered with a comment - `/azp run <pipeline-name>`
 * `runtime`, and `runtime-staging`, run jobs only when relevant paths change. And for `EAT`, and `AOT`, only smoke tests are run.
 * And when `runtime-wasm` is triggered manually, it runs *all* the wasm jobs completely

 | .                 | runtime               | runtime-staging         | runtime-extra-platforms(manual only) | runtime-wasm (manual only) |
 | ----------------- | --------------------  | ---------------         | ------------------------------------ | -------                    |
 | libtests          | linux: all,   only-pc | windows: all,   only-pc | linux+windows: all, only-pc          | linux+windows: all, always |
 | libtests eat      | linux: smoke, only-pc | -                       | linux:         all, only-pc          | linux:         all, always |
 | libtests aot      | linux: smoke, only-pc | windows: smoke, only-pc | linux+windows: all, only-pc          | linux+windows: all, always |
 |                   |                       |                         |                                      |                            |
 | Wasm.Build.Tests  | linux:        only-pc | windows:        only-pc | linux+windows: only-pc               | linux+windows              |
 | Debugger tests    | -                     | linux+windows:  only-pc | linux+windows: only-pc               | linux+windows              |
 | Runtime tests     | linux:        only-pc | -                       | linux: only-pc                       | linux                      |

 ## Rolling build (twice a day):

 * `runtime`, and `runtime-staging`, run all the wasm jobs unconditionally, but `EAT`, and `AOT` still run only smoke tests.
 * `runtime-extra-platforms` also runs by default. And it runs only the cases not covered by the above two pipelines.

 * jobs w/o `only-pc` are always run

 | .                 | runtime                   | runtime-staging       | runtime-extra-platforms (always run) | runtime-wasm (manual only) |
 | ----------------- | -------------             | ---------------       | ------------------------------------ | ------                     |
 | libtests          | linux: all(v8/chr)        | windows: all          | none                                 | N/A                        |
 | libtests eat      | linux: smoke              | -                     | linux: all                           |                            |
 | libtests aot      | linux: smoke              | windows: smoke        | linux+windows: all                   |                            |
 |                   |                           |                       |                                      |                            |
 | Wasm.Build.Tests  | linux: always             | windows: always       | none                                 |                            |
 | Debugger tests    | -                         | linux+windows: always | none                                 |                            |
 | Runtime tests     | linux: always             | -                     | none                                 |                            |

### CODE REVIEW

**Comment by radical** [comment]
Two ways to do this:

1. For rolling builds, build all the wasm jobs in runtime+runtime-staging
    - this is what I tried to implement here, but I didn't realize that `runSmokeTestsOnly` will be `true` for runtime+runtime-staging, even on rolling builds. So, we would need to make it true only on PRs.

2. For rolling builds, don't run any wasm jobs in runtime+runtime-staging, instead run all of them in runtime-extra-platforms

Which one makes sense?

**Comment by radical** [comment]
@safern any thoughts on https://github.com/dotnet/runtime/pull/65035#issuecomment-1033307306 ?

**Comment by safern** [comment]
> @safern any thoughts on https://github.com/dotnet/runtime/pull/65035#issuecomment-1033307306 ?

This goes against the goal that we had with adding `runtime-extra-platforms` pipeline, the goal and intent is that `runtime` PR builds are the same as `runtime` rolling builds, so if we run smoke tests on PRs, the rolling build should run the same set of tests. This helps CI Council and infrastructure team evaluate and investigate pass rate for the pipeline and know in what state our PRs are in. 

I would try and do the same for `runtime-extra-platforms` as rolling builds is our only input of data on how stable a pipeline is (this is because PR builds can fail due to a change on the PR itself, not because the pipeline is broken in that branch).

**Comment by radical** [comment]
Need to confirm what `runtime-extra-platforms`, and `runtime-wasm` would run on the PR.

File: eng/pipelines/common/templates/wasm-library-tests.yml [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+#
+# Build for Browser/wasm and test it on v8, browser, and nodejs
+#
+- template: /eng/pipelines/common/platform-matrix.yml
+  parameters:
+    jobTemplate: /eng/pipelines/common/global-build-job.yml
+    helixQueuesTemplate: /eng/pipelines/libraries/helix-queues-setup.yml
+    buildConfig: Release
+    runtimeFlavor: mono
```
Comment (melvinmezzy) [inline eng/pipelines/common/templates/wasm-library-tests.yml:21]: For my information

File: eng/pipelines/common/templates/wasm-library-tests.yml [hunk 2]
```diff
@@ -0,0 +1,51 @@
+parameters:
+  alwaysRun: false
```
Comment (melvinmezzy) [inline eng/pipelines/common/templates/wasm-library-tests.yml:2]: True

**Review by melvinmezzy — APPROVED** [review]
Feedback

**Comment by radical** [comment]
Merged `main` to resolve new conflicts.

**Review by lewing — APPROVED** [review]
Any obections?

**Comment by radical** [comment]
This is essentially moving the wasm jobs to separate template files, and then using them in the original locations.

**Comment by radical** [comment]
Wasm.Build.Tests haven't run in more than 14days now, this fixes the case.

**Pull request merged**